### PR TITLE
readme : add link to Autopen under UIs

### PR DESCRIPTION
**Pull request by blackhole89**
Sorry that after all the months of AWOL this triviality is all I come back with, but I'd be grateful if my little project could be added to the README list.

[Autopen](https://github.com/blackhole89/autopen) is a graphical text editor that uses llama.cpp to tokenize the buffer on the fly, score the buffer, visualise token logits and allow you to switch back and forth between different possible completions at any point. There's a demo video [here](https://www.youtube.com/watch?v=GXWZPpVI0zU). I hope the criteria for inclusion are met, as I'm stating the dependency prominently.

(I think I might also have found some subtle issues (of numerical stability, snapshot save/restore, and/or the CUDA kernels) in the process of working on it, where different sequences of decode calls that *should* give identical results don't, but I'm still working on isolating the exact conditions in a way that is actionable.)

### CODE REVIEW

**Review by slaren — APPROVED** [review]

**Comment by ggerganov** [comment]
> (I think I might also have found some subtle issues (of numerical stability, snapshot save/restore, and/or the CUDA kernels) in the process of working on it, where different sequences of decode calls that should give identical results don't, but I'm still working on isolating the exact conditions in a way that is actionable.)

Indeed there are some subtle effects that lead to differences when processing the same input with different batch sizes (see for example https://github.com/ggerganov/llama.cpp/pull/7745). Not clear how to make this fully deterministic. There are at least 2 points in the computation that are problematic in this regard:

- The `max` computation over the sequence length in the `softmax` operator of the attention
- The KQ*V reduction over the sequence length at the end of the attention

**Comment by blackhole89** [comment]
Ah, good to know that it's a known problem. I've observed discrepancies on the order of 0.5 logits (usually less), though my use case may be a bit pathological since I frequently reevaluate the same run of ~10 tokens (the window between two snapshots, with default params, in my program) batched up differently, in a context where the ordering of two choices suddenly flipping is quite conspicuous.
